slapo/schedule.py

```diff
@@ -407,6 +409,44 @@ def validate_sync_op(mode, sync_op_or_fn, axis=None):
                 if mode == "fwd_post" and output_type == "partition":
                     raise ValueError("Cannot all-reduce a partition output")
 
+        def decouple_bias_from_linear(sch, sync_fn):
+            param = sch.mod.get_parameter("weight")
+            # Replace the linear module
+            with init_empty_weights(
+                enable=(sch.mod.weight.device == torch.device("meta"))
+            ):
+                new_mod = LinearWithSeparateBias(
+                    sch.mod.weight.shape[1],
+                    sch.mod.weight.shape[0],
+                    sync_fn,
+                    sch.mod.weight.device,
+                    sch.mod.weight.dtype,
+                )
+                new_mod.in_features = sch.mod.in_features
+                new_mod.out_features = sch.mod.out_features
+                sch.replace(new_mod)
+            # Test if the original weight is tied to other modules
+            new_tensor = new_mod.weight.data
+            if param in sch.metadata.tie_weights:
+                if id(sch.metadata.tie_weights[param]) != id(param):
+                    # This parameter is tied to another parameter.
+                    # We directly register the sharded parameter to the module
+                    # to keep them tied.
+                    if new_tensor.shape != sch.metadata.tie_weights[param].shape:
+                        raise RuntimeError(
+                            f"Parameter weight in {sch.path} is tied, "
+                            "but they have different sharded shapes: "
+                            f"{new_tensor.shape} vs "
+                            f"{sch.metadata.tie_weights[param].shape}"
+                        )
+                    new_param = sch.metadata.tie_weights[param]
+                else:
+                    new_param = new_mod.weight
+                    self.metadata.tie_weights[param] = new_param
+            else:
+                new_param = new_mod.weight
```

Comment:
Could we have a utility function for this logic so that we could just call it when needed?

Reply:
Added